### ISSUE TITLE
Conditional Module imports (e.g. for development only)

### ISSUE THREAD

**Issue by jbjhjm** [angular/angular#36243]
# 🚀 Conditional Module imports (e.g. for development-only modules)

### Relevant Package

 This feature request is for @angular/core / ivy

### Description
View Engine allowed us to pass a module conditionally, which was useful as some modules should only be imported in development. Example:

```ts
@NgModule({
  imports: [
    process.env.NODE_ENV === 'development' ? NgxsReduxDevtoolsPluginModule.forRoot({})  : undefined
  ]
})
```

Using Ivy, the above solution will fail when building with production target.
Running ng serve xxx --prod leads to these errors:
CLI will show lots of misguiding errors (telling me component xxx is not a known element).
Browser console will show "ASSERTION ERROR: Type passed in is not NgModuleType, it does not have 'ɵmod' property."

I thought Ivy might not like the undefined value, so I tried this as a workaround:

```ts
@NgModule()
export class NoopModule {}

@NgModule({
  imports: [
    process.env.NODE_ENV === 'development' ? NgxsReduxDevtoolsPluginModule.forRoot({})  : NoopModule 
  ]
})
```

... which lead to the same errors. 

### Describe the solution you'd like

Does Ivy support this by now, in way I don't know?
if not, I suggest to allow to pass undefined to module imports.

**Comment by BramMeerten** [comment]
I have the same problem. The following worked in angular 8.2.14:
```
@NgModule({
    imports: [
        // ...
        areAnimationsDisabled() ? NoopAnimationsModule : BrowserAnimationsModule
    ]
})
```

But I get these errors after I upgrade to angular 9.1.0:
` error NG1010: Value at position 1 in the NgModule.imports of AppModule is not a reference: [object Object]`

**Comment by AliYusuf95** [comment]
I face the same issue, to make the project works I have to disable ivy `"enableIvy": false`


```log
.... - error NG1010: Value at position 28 in the NgModule.imports of ChildModule is not a reference: [object Object]

202     imports: [
                 ~
203         MatDialogModule,
    ~~~~~~~~~~~~~~~~~~~~~~~~
... 
257         MatSidenavModule,
    ~~~~~~~~~~~~~~~~~~~~~~~~~
258     ],
    ~~~~~
.... - error NG6002: Appears in the NgModule.imports of ParentModule, but could not be resolved to an NgModule class.
...
```

This is my module
```ts
@NgModule({
    imports: [
        MatDialogModule,
        CommonModule,
        ...
       // This conditional import cause the issue
        environment.isMockEnabled
            ? HttpClientInMemoryWebApiModule.forFeature(FakeApiService, {
                  passThruUnknownUrl: true,
                  dataEncapsulation: false,
              })
            : [],
        ...
        MatSidenavModule,
    ],
    declarations: [ChildComponent],
})
export class ChildModule {}
```

**Comment by alxhub** [comment]
(I'm struggling to categorize this correctly, hence the label dance)

Angular (Ivy) today requires that certain expressions (e.g. the `imports` of an `NgModule`) be statically understandable. Unfortunately `process.env` is not, there's no way for our compiler to know at _build time_ what value this will take when executed.

The CLI has a feature (`environments/environment.ts`) for this purpose - it allows you to vary the value of constants based on the current environment, which is controlled (I think) from `angular.json`. This feature works with our static evaluation, so you can do things like:

```typescript
@NgModule({
  imports: [
    environment.flag ? SomeModule : [],
  ],
})
export class ConfigSpecificModule {}
```

**Comment by alxhub** [comment]
@AliYusuf95 as for your example, I'd have to see a reproduction to know why it's not working in your case (please file a separate issue if you'd like to figure it out). My suspicion is that something inside the environment file is not analyzable.

**Comment by AliYusuf95** [comment]
@alxhub I used `val-loader` in my environment files to [add the ability to use dotenv file](https://codingsans.com/blog/configure-frontend-projects-with-dotenv)

So I have `environment-loader.js` file with:
```js
const convict = require('convict');
const dotenv = require('dotenv');

dotenv.config();

const config = convict({ ...... });

config.validate({ allowed: 'strict' });

module.exports = () => {
    return {
        code: 'module.exports = ' + JSON.stringify(config.getProperties()),
    };
};
```

And `environment.interface.ts` file:
```ts

import * as loadedEnv from '!val-loader!./environment-loader';

export interface IEnvironment {
    production: boolean;
    ....
}

export const env = loadedEnv as IEnvironment;
export function buildEnvironment(
    environment: Partial<IEnvironment>,
): IEnvironment {
    return Object.assign({}, env, environment);
}
```

And finally `environment.ts` file:
```ts
import { buildEnvironment } from './environment.interface';

export const environment = buildEnvironment({
    production: false,
    .....
});

```

At build time `environment-loader.js` file is converted to basic module file `module.exports = ...`, does that what cased the issue?

**Comment by jbjhjm** [comment]
Ok I think the issue gets transparent now. 
Ivy build process will only accept static boolean comparisions and it cannot resolve  process.env.NODE_ENV or any type of dynamic configuration, like val-loader used by @AliYusuf95 or my setup which uses a custom working similarly system.

I see two different concepts here, for one static, compile-time variables and on the other hand dynamic, possibly complex application settings which should be evaluated when running the application.
So what we are looking for is more like a flag determining if we are in development environment or not.

For process.env.NODE_ENV, I'm used to seeing it be replaced by a static value during webpacking. I wonder why this does not seem to happen anymore.

### A (failed) attempt to work around this issue

As I already am using a webpack-config loader I was hoping I could use
```ts
new webpack.DefinePlugin({
	'process.env.NODE_ENV': ...
	// or maybe better to use
	'process.env.IN_DEVELOPMENT_ONLY': ...
}),
```
As process.env.NODE_ENV is not populated by cli, one would have to pass that in by command, or use the resolved architect configuration. Which ends with another dilemma: The architect does not categorize into "development" or "production", so how should one decide if the requested task is for development or production?

Perhaps one could check the value of one of the typical production settings.
However this would lead to very intransparent behavior.

Or one could add another fileReplacements entry providing a static, build-time configuration.
This might be more transparent, but not efficient and would clutter the workspace with more files.

`{
              "fileReplacements": [],
              "optimization": true,
              "outputHashing": "none",
              "sourceMap": true,
              "extractCss": true,
              "namedChunks": false,
              "aot": true,
              "extractLicenses": true,
              "vendorChunk": false,
              "buildOptimizer": true,
 }`


To sum up, I think there's a need for static "flag" options as well as runtime/environment configuration, which are incompatible with each other unfortunately.

**Comment by jbjhjm** [comment]
ok there is one __important__ thing!!

One MUST NOT pass undefined into imports list under any circumstances.

This will fail during production build with Error `Unexpected value 'undefined' imported by the module ...`
```ts
imports: [
	environment.BUILD_MODE==='development' ? MyDevOnlyModule : undefined
],
```

But if you pass in an empty array as @alxhub does in his example above, it will work.
```ts
imports: [
	environment.BUILD_MODE==='development' ? MyDevOnlyModule : []
],
```

So examples and issues mentioned above might be due to dynamic environment. But the environment may also actually be analyzed correctly and the "undefined" value might be source of the problem.

**Comment by LanderBeeuwsaert** [comment]
We have the same request/requirement where we want to be able to disable animations in most of our runs for e2e testing. And we pass this in a non-statical way:
shouldSpeedUpE2e() ? NoopAnimationsModule : BrowserAnimationsModule

**Comment by JoostK** [comment]
@LanderBeeuwsaert Angular 12 supports that particular use-case using [the new `BrowserAnimationsModule.withConfig` API](https://rc.angular.io/api/platform-browser/animations/BrowserAnimationsModule#withconfig).

**Comment by JoostK** [comment]
This pattern has never been supported in AOT mode and it may have worked for the wrong reasons with the ViewEngine compiler. A conditional import requires that the condition is known at compile time and NodeJS style conditions like `process.env.NODE_ENV` are not known to the compiler.

The CLI supports conditional NgModule imports in an application using environments files. The environment files get replaced at compile time such that the AOT compiler sees the desired value when statically evaluating the condition. If any of the ternary's branches should not include any NgModule, then `[]` should be used instead of `undefined` (as the latter is an invalid NgModule import, whereas the empty array would indicate that zero NgModules have to be imported.)
